sys/page_revisions.js

```diff
@@ -80,6 +83,190 @@ PRS.prototype.getTableSchema = function() {
+            redirect: 'string',
+            page_deleted: 'int'
+        },
+        index: [
+            { attribute: 'title', type: 'hash' },
+            { attribute: 'rev', type: 'range', order: 'asc' },
+            { attribute: 'page_deleted', type: 'static' }
+        ]
+    };
+};
+
+/**
+ * Returns the suppression table URI for a given domain
+ * @param {string} domain the domain
+ * @returns {URI} suppression table URI
+ */
+PRS.prototype.restrictionsTableURI = function(domain) {
+    return new URI([domain, 'sys', 'table', this.restrictionsTableName, '']);
+};
+
+PRS.prototype.getRestrictions = function(hyper, req) {
+    var self = this;
+    var rp = req.params;
+    var attributes = { title: rp.title };
+    if (rp.revision) {
+        attributes.rev = {
+            ge: rp.revision
+        };
+    }
+    return hyper.get({
+        uri: self.restrictionsTableURI(rp.domain),
+        body: {
+            table: self.restrictionsTableName,
+            attributes: attributes,
+            limit: 1
+        }
+    })
+    .then(function(res) {
+        // Remove possible revision restrictions as here we just need
+        // the page deletion info
+        var restrictions = res.body && res.body.items && res.body.items[0] || null;
+        if (restrictions) {
+            if (restrictions.rev !== rp.revision && restrictions.restrictions) {
+                restrictions.restrictions = [];
+            }
+            res.body = restrictions;
+        } else {
+            res.body = null;
+        }
+        return res;
+    });
+};
+
+/**
+ * Update restrictions for a title & revision. Used primarily by the parsoid
+ * module to update redirects on save.
+ */
+PRS.prototype.postRestrictions = function(hyper, req) {
+    var rp = req.params;
+    // Validate the request body
+    var body = req.body;
+    if (!body || !(body.restrictions || body.redirect)) {
+        throw new HTTPError({
+            status: 400,
+            body: {
+                type: 'bad_request',
+                description: 'Expected restrictions or redirect in the POST body.',
+            }
+        });
+    }
+    var revision = {
+        title: rp.title,
+        rev: rp.revision,
+    };
+    Object.assign(revision, body);
+    return this.storeRestrictions(hyper, req, revision);
+};
+
+// Const TimeUuid to enable partial restriction updates.
+var tidNode = new Buffer([0x01, 0x23, 0x45, 0x67, 0x89, 0xab]);
+var tidClock = new Buffer([0x12, 0x34]);
+var restrictionsTid = new TimeUuid(new Date(0), 0, tidNode, tidClock).toString();
+
+PRS.prototype.storeRestrictions = function(hyper, req, revision) {
+    var self = this;
+    var rp = req.params;
+    // Do not even define attributes we don't want to overwrite.
+    var restrictionObject = {};
+    if (revision.restrictions && revision.restrictions.length) {
+        restrictionObject.restrictions = revision.restrictions;
+    }
+    if (revision.redirect !== true && revision.redirect) {
```

Comment:
Oh, so the `revision` parameter here is not the revision we get from MW API.. A bit misleading.